README.md
<p align="center">
 <a href="https://www.cyberghostvpn.com/"> <img src="https://raw.githubusercontent.com/tmcphee/cyberghostvpn/main/.img/CyberGhost-Logo-Header.png"></a>
</p>

# CyberGhost VPN
![GitHub Workflow Status](https://img.shields.io/github/actions/workflow/status/tmcphee/cyberghostvpn/docker-image.yml?style=for-the-badge) ![GitHub release (latest by date)](https://img.shields.io/github/v/release/tmcphee/cyberghostvpn?style=for-the-badge)

This is a WireGuard client docker that uses the CyberGhost CLI. It allows routing containers traffic through WireGuard.

[Docker Image](https://hub.docker.com/r/tmcphee/cyberghostvpn)
###### Ubuntu 20.04 | CyberGhost 1.3.4
## What is WireGuard?

WireGuard® is an extremely simple yet fast and modern VPN that utilizes state-of-the-art cryptography. It aims to be faster, simpler, leaner, and more useful than IPsec, while avoiding the massive headache. It intends to be considerably more performant than OpenVPN. WireGuard is designed as a general-purpose VPN for running on embedded interfaces and super computers alike, fit for many different circumstances.

## How to use this image
Start the image using optional environment variables shown below. The end-user must supply a volume for local storage of the CyberGhost auth and token files. Supplied DNS is optional to avoid using ISP DNS during the initial connection. Ensure to run the image in privileged mode.
```
docker run 
   -d 
   --name='cyberghostvpn'
   --net='bridge'
   --privileged=true
   --cap-add=NET_ADMIN 
   -e TZ="America/New_York"
   -e 'ACC'='example@gmail.com'
   -e 'PASS'='mypassword'
   -e 'COUNTRY'='US'
   -e 'NETWORK'='192.168.1.0/24'
   -e 'WHITELISTPORTS'='9090,8080'
   -p 9090:9090
   -p 8080:8080
   -v '/local/path/to/config':'/home/root/.cyberghost:rw'
```

Other containers can connect to this image by using its network connection.
`--net=container:cyberghostvpn`
```
docker run -d --net=container:cyberghostvpn other-container
```
Note: If the other containers have exposed ports for example a WEBUI. Forward that port in the cyberghostvpn image, add the port to WHITELISTPORTS environment variable, and set your local LAN using NETWORK environment variable. See [Environment variables](https://github.com/tmcphee/cyberghostvpn#environment-variables) below for details. 

## Selecting a country

Add an environment variable called `COUNTRY` and set to the desired country. 
Examples:
- `United States` COUNTRY=US
- `Canada`        COUNTRY=CA

See [CyberGhost selecting a country or single server](https://support.cyberghostvpn.com/hc/en-us/articles/360020673194--How-to-select-a-country-or-single-server-with-CyberGhost-on-Linux) for more details

## Custom DNS / NAMESERVER
Add an environment variable called `NAMESERVER` and set to the desired DNS. 
Examples:
- Cloudflare 1.1.1.1
- Google 8.8.8.8

This image will use CyberGhost Smart DNS if no Nameserver is provided. Automatic Smart DNS for countries US, NL, JP and GB. Default is CloudFlare [1.1.1.1] for all other countries.

## How to login
Login by providing the ACC and PASS environment variables
```
           -e ACC=example@gmail.com
           -e PASS=mypassword
```

## How to access ports locally
Access ports [webUI] by providing the NETWORK and WHITELISTPORTS environment variables. Where NETWORK is the user’s network and WHITELISTPORTS is the ports the user wants to expose. 
```
           -e NETWORK=192.168.1.0/24
           -e WHITELISTPORTS=9090,8080
           -p 9090:9090
           -p 8080:8080
```

## HTTP Proxy Service
Connect devices on the same network with HTTP proxy. Allowing to use the same IP address as the VPN client.
Proxy will disconnect if VPN conection is starting or lost. 
Proxy Port: 3128
```
           -e PROXY=True
           -p 3128:3128
```
Connect devices with the HOST IP and proxy port 3128

## Environment variables

- `NETWORK` - Adds a route to the local network once the VPN is connected. CIDR networks [192.168.1.0/24]
- `WHITELISTPORTS` - Allow access to listed ports when VPN is connected. Delimited by comma [8080,8081,9000]
- `ACC` - CyberGhost username - Used for login
- `PASS` - CyberGhost password - Used for login
- `COUNTRY` - Destination Country - See [CyberGhost Connect to a country](https://support.cyberghostvpn.com/hc/en-us/articles/360020673194--How-to-select-a-country-or-single-server-with-CyberGhost-on-Linux#h_01EJDGC9TZDW38J9FKNFPE6MBE)
- `ARGS` - All additional arguments [Examples: "--torrent" "--traffic" "--streaming 'Netflix US'"]
- `NAMESERVER` - Custom Nameserver/DNS [Examples: Cloudflare 1.1.1.1, Google 8.8.8.8]
- `PROTOCOL` - Choose between WireGuard or OpenVPN [wireguard, openvpn]. Default WireGuard
- `FIREWALL` - Optional disable firewall. [FIREWALL=False]. Default True
- `PROXY` - Optional enable proxy. [PROXY=True]. Default False

## Firewall
This image has a custom built-in firewall. On initial start, all traffic is blocked except CyberGhost API IP and Local DNS for resolve. After VPN is connected Local DNS is blocked on Port 53. For first time use the firewall will go through a setup phase to include whitelisted ports where the firewall will be inactive. 

See the firewall section located in start.sh for details. 

## Troubleshooting

Docker runs, but WireGuard does not connect or gives an error
- Try deleting the config.ini file located in your mapped config folder. This file is the login token for CyberGhost and may be expired. 


## Disclaimer
This project was developed independently for personal use. CyberGhost has no affiliation, nor has control over the content or availability of this project.


ChangeLog.txt

#1.2.2
	- Add HTTP Proxy Service
	- Proxy Toggle ENV variable
#1.2.1
	- Update Ubuntu Version to 20.04
	- Suppressed Commands
#1.1.4
	- Ping to check Internet reachability every 30 minutes
	- Error catching for incorrect login auth [config.ini]
	- Changed default DNS to CloudFlare 1.1.1.1
	- Error codes for tracing
	- No cache headers added to IP and Location check
	- Firewall Toggle ENV variable
	- Protocal Select ENV Variable [WireGuard, OpenVPN]
	
#1.1.3
	- Run.sh factored into main script [start.sh] for speed improvements
	- Variable printouts for Debugging
	- Removed unused imports iptables, net-tools, and ifupdown
	
#1.1.1
	- Update CyberGhost CLI Version to 1.3.4
	- New variables COUNTRY, ARGS, and NAMESERVER
	- Updated Run.sh with new DNS sources, error catching, and tweaks
	- Updated Start.sh to show VPN connection info after successful connection
	
#1.0.1
	- Added login by environment variables
	- Fix for routing

#1.0.0
	- Initial Realease


run.sh
#!/bin/bash
#VER2.1
	config_ini=/home/root/.cyberghost/config.ini
	if [ -f "$config_ini" ]; then
	
		# Check if country is set. Default to US
		if ! [ -n "$COUNTRY" ]; then
			echo "Country variable not set. Defaulting to US"
			export COUNTRY="US"
		fi
			
		#Launch and connect to CyberGhost VPN
		sudo cyberghostvpn --connect --country-code $COUNTRY --wireguard $ARGS
		
		# Add CyberGhost nameserver to resolv for DNS
		# Add Nameserver via env variable $NAMESERVER
		if [ -n "$NAMESERVER" ]; then
			echo 'nameserver ' $NAMESERVER > /etc/resolv.conf
		else
			# SMART DNS
			# This will switch baised on country selected
			# https://support.cyberghostvpn.com/hc/en-us/articles/360012002360
			case "$COUNTRY" in
				"NL") echo 'nameserver 75.2.43.210' > /etc/resolv.conf
				;;
				"GB") echo 'nameserver 75.2.79.213' > /etc/resolv.conf
				;;
				"JP") echo 'nameserver 76.223.64.81' > /etc/resolv.conf
				;;
				"DE") echo 'nameserver 13.248.182.241' > /etc/resolv.conf
				;;
				"US") echo 'nameserver 99.83.181.72' > /etc/resolv.conf
				;;
				*) echo 'nameserver 99.83.181.72' > /etc/resolv.conf
				;;
		esac
		fi
	fi
	


start.sh
#!/bin/bash

	config_ini=/home/root/.cyberghost/config.ini #CyberGhost Auth token

	enable_dns_port () {
		echo "Allowing PORT 53 - IN/OUT"	
		sudo ufw allow out 53 > /dev/null 2>&1 #Allow port 53 on all interface for initial VPN connection
		sudo ufw allow in 53 > /dev/null 2>&1
	}
	
	disable_dns_port () {
		echo "Blocking PORT 53 - IN/OUT"
		sudo ufw delete allow out 53 > /dev/null 2>&1 #Remove Local DNS Port to prevent leaks
		sudo ufw delete allow in 53 > /dev/null 2>&1
	}
	
	startup () {
		echo "CyberGhostVPN - Docker Edition"
		echo "----------------------------------------------------------"
		echo "	Created By: Tyler McPhee"
		echo "	GitHub: https://github.com/tmcphee/cyberghostvpn"
		echo "	DockerHub: https://hub.docker.com/r/tmcphee/cyberghostvpn"
		echo "	"
		echo "	Ubuntu:${linux_version} | CyberGhost:${cyberghost_version} | ${script_version}"
		echo "----------------------------------------------------------"
		
		echo "**************User Defined Variables**************"
		
		if [ -n "$ACC" ]; then
			echo "	ACC: [PASSED - NOT SHOWN]"
		fi
		if [ -n "$PASS" ]; then
			echo "	PASS: [PASSED - NOT SHOWN]"
		fi
		
		if [ -n "$COUNTRY" ]; then
			echo "	COUNTRY: ${COUNTRY}"
		fi
		if [ -n "$NETWORK" ]; then
			echo "	NETWORK: ${NETWORK}"
		fi
		if [ -n "$WHITELISTPORTS" ]; then
			echo "	WHITELISTPORTS: ${WHITELISTPORTS}"
		fi
		if [ -n "$ARGS" ]; then
			echo "	ARGS: ${ARGS}"
		fi
		if [ -n "$NAMESERVER" ]; then
			echo "	NAMESERVER: ${NAMESERVER}"
		fi
		if [ -n "$PROTOCOL" ]; then
			echo "	PROTOCOL: ${PROTOCOL}"
		fi
		if [ -n "$PROXY" ]; then
			echo "	PROXY: ${PROXY}"
		fi

		echo "**************************************************"
		
	}
	
	ip_stats () {
		str="$(cat /etc/resolv.conf)"
		value=${str#* }
		
		echo "***********CyberGhost Connection Info***********"
		echo "	IP: ""$(curl -s https://ipinfo.io/ip -H "Cache-Control: no-cache, no-store, must-revalidate")"
		echo "	CITY: ""$(curl -s https://ipinfo.io/city -H "Cache-Control: no-cache, no-store, must-revalidate")"
		echo "	REGION: ""$(curl -s https://ipinfo.io/region -H "Cache-Control: no-cache, no-store, must-revalidate")"
		echo "	COUNTRY: ""$(curl -s https://ipinfo.io/country -H "Cache-Control: no-cache, no-store, must-revalidate")"
		echo "	DNS: ${value}"
		echo "************************************************"
	}
	
	#Originated from Run.sh. Migrated for speed improvements
	cyberghost_start () {
		#Stop Proxy service to prevent dns leaks
		if [ -n "${PROXY}" ]; then
			if [ "${PROXY}" == "True" ]; then
				echo Stopping HTTP Proxy...
				sudo systemctl disable squid
				sudo systemctl stop squid
				sudo systemctl status squid
			fi
		fi
		enable_dns_port
		#Check for CyberGhost Auth file
		if [ -f "$config_ini" ]; then
	
			# Check if country is set. Default to US
			if ! [ -n "$COUNTRY" ]; then
				echo "Country variable not set. Defaulting to US"
				export COUNTRY="US"
			fi
				
			# Check if protocol is set. Default WireGuard
			if ! [ -n "$PROTOCOL" ]; then
				export PROTOCOL="wireguard"
			fi
				
			#Launch and connect to CyberGhost VPN
			sudo cyberghostvpn --connect --country-code "$COUNTRY" --"$PROTOCOL" "$ARGS"
			
			# Add CyberGhost nameserver to resolv for DNS
			# Add Nameserver via env variable $NAMESERVER
			if [ -n "$NAMESERVER" ]; then
				echo 'nameserver ' "$NAMESERVER" > /etc/resolv.conf
			else
				# SMART DNS
				# This will switch baised on country selected
				# https://support.cyberghostvpn.com/hc/en-us/articles/360012002360
				case "$COUNTRY" in
					"NL") echo 'nameserver 75.2.43.210' > /etc/resolv.conf
					;;
					"GB") echo 'nameserver 75.2.79.213' > /etc/resolv.conf
					;;
					"JP") echo 'nameserver 76.223.64.81' > /etc/resolv.conf
					;;
					"DE") echo 'nameserver 13.248.182.241' > /etc/resolv.conf
					;;
					"US") echo 'nameserver 99.83.181.72' > /etc/resolv.conf
					;;
					*) echo 'nameserver 1.1.1.1' > /etc/resolv.conf
					;;
			esac
			fi
		fi
		disable_dns_port
		#Enable Proxy service
		if [ -n "${PROXY}" ]; then
			if [ "${PROXY}" == "True" ]; then
				echo Starting HTTP Proxy...
				sudo systemctl enable squid
				sudo systemctl start squid
				sudo systemctl status squid
			fi
		fi
		ip_stats
	}
	
	#Check if the site is reachable
	check_up() {
		ping -c1 "1.1.1.1" > /dev/null 2>&1 #Ping CloudFlare
		if [ $? -eq 0 ]; then
			return 0
		fi
		return 1
	}
	if ! [ -n "$FIREWALL" ]; then
		export FIREWALL="True"
	fi
	
	startup
	
	#Check if CyberGhost CLI is installed. If not install it
	FILE=/usr/local/cyberghost/uninstall.sh
	if [ ! -f "$FILE" ]; then
		echo "CyberGhost CLI not installed. Installing..."
		bash /install.sh
		echo "Installed"
	fi
	
	#Run Firewall if Enabled. Default Enabled
	sysctl -w net.ipv6.conf.all.disable_ipv6=1 #Disable IPV6
	sysctl -w net.ipv6.conf.default.disable_ipv6=1
	sysctl -w net.ipv6.conf.lo.disable_ipv6=1
	sysctl -w net.ipv6.conf.eth0.disable_ipv6=1
	sysctl -w net.ipv4.ip_forward=1
	
	
	if [[ ${FIREWALL,,} == *"true"* ]]; then
		sudo ufw enable #Start Firewall

		FIREWALL_FILE=/.FIREWALL.cg
		if [ ! -f "$FIREWALL_FILE" ]; then
			echo "Initiating Firewall First Time Setup..."
				
			sudo ufw disable #Stop Firewall
			export CYBERGHOST_API_IP=$(getent ahostsv4 v2-api.cyberghostvpn.com | grep STREAM | head -n 1 | cut -d ' ' -f 1)
			sudo ufw default deny outgoing > /dev/null 2>&1							#Deny All traffic by default on all interfaces
			sudo ufw default deny incoming > /dev/null 2>&1
			sudo ufw allow out on cyberghost from any to any > /dev/null 2>&1 		#Allow All over cyberghost interface
			sudo ufw allow in on cyberghost from any to any > /dev/null 2>&1
			sudo ufw allow in 1337 > /dev/null 2>&1 								#Allow port 1337 for CyberGhost Communication
			sudo ufw allow out 1337 > /dev/null 2>&1
			sudo ufw allow in 891 > /dev/null 2>&1 									#Allow port 1194 for CyberGhost OpenVPN Communication
			sudo ufw allow out 819 > /dev/null 2>&1
			sudo ufw allow out from any to "$CYBERGHOST_API_IP" > /dev/null 2>&1 	#Allow v2-api.cyberghostvpn.com [104.20.0.14] IP for connection
			sudo ufw allow in from "$CYBERGHOST_API_IP" to any > /dev/null 2>&1
			
			#Allow all ports in WHITELISTPORTS ENV [Seperate by ',']
			if [ -n "${WHITELISTPORTS}" ]; then
				echo "Setting Whitelisted Ports..."
				IFS=',' read -a array <<< "$WHITELISTPORTS"
				for i in "${array[@]}"
				do
				   echo "Whitelisting Port:" "$i"
				   sudo ufw allow "$i" > /dev/null 2>&1
				done
			fi
			
			sudo ufw enable #Start Firewall
			echo "Firewall Setup Complete"	
			echo 'FIREWALL ACTIVE WHEN FILE EXISTS' > .FIREWALL.cg
		fi
	else
		sudo ufw disable #Stop Firewall
	fi
	
	#Login to account if config not exist
	if [ ! -f "$config_ini" ]; then
		echo "Logging into CyberGhost..."
		
		#Check for CyberGhost Credentials and Login
		if [ -n "$ACC" ] && [ -n "$PASS" ]; then
			enable_dns_port
			expect /auth.sh
			disable_dns_port
		else
			echo "[E1] Can't Login. User didn't provide login credentials. Set the ACC and PASS ENV variables and try again." 
			exit
		fi
	else
		#Verify the config.ini has successfully created the Account and assigned a Device
		echo "Verifying Login Auth..."
		if ! grep -q '[Device]' $config_ini; then
			echo "Failed"
			rm "$config_ini"
			echo "Logging into CyberGhost..."
			enable_dns_port
			expect /auth.sh
			disable_dns_port
		else
			echo "Passed"
		fi
	fi
	
	if [ -n "${NETWORK}" ]; then
		echo "Adding network route..."
		export LOCAL_GATEWAY=$(ip r | awk '/^def/{print $3}') # Get local Gateway
		ip route add "$NETWORK" via "$LOCAL_GATEWAY" dev eth0 #Enable access to local lan
		echo "$NETWORK" "routed to" "$LOCAL_GATEWAY" "on eth0"
	fi
	
	
	if [ -n "${PROXY}" ]; then
		if [ "${PROXY}" == "True" ]; then
			echo "Seting up HTTP proxy on port 3128..."
			sudo ufw allow in 3128 > /dev/null 2>&1 #Enable Proxy Port
			sudo ufw allow out 3128 > /dev/null 2>&1
		else
			echo "Disabling HTTP proxy..."
			sudo ufw deny in 3128 > /dev/null 2>&1 #Disable Proxy Port
			sudo ufw deny out 3128 > /dev/null 2>&1
		fi
	else
		echo "Disabling HTTP proxy..."
		sudo ufw deny in 3128 > /dev/null 2>&1 #Disable Proxy Port
		sudo ufw deny out 3128 > /dev/null 2>&1
	fi
	
	#WIREGUARD START AND WATCH
	cyberghost_start
	t_hr="$(date -u --date="+30 minutes" +%H)" #Next time to check internet is reachable
	t_min="$(date -u --date="+30 minutes" +%M)"
	while true #Watch if Connection is lost then reconnect
	do
		sleep 30
		if [[ $(sudo cyberghostvpn --status | grep 'No VPN connections found.' | wc -l) = "1" ]]; then
			echo '[E2] VPN Connection Lost - Attempting to reconnect....'
			cyberghost_start	
		fi
		
		#Every 30 Minutes ping CloudFlare to check internet reachability 
		if [ "$(date +%H)" = "$t_hr" ] && [ "$(date +%M)" = "$t_min" ]; then
			if ! check_up; then
				echo '[E3] Internet not reachable - Restarting VPN...'
				sudo cyberghostvpn --stop
				cyberghost_start
				t_hr="$(date -u --date="+30 minutes" +%H)" #Next time to check internet is reachable
				t_min="$(date -u --date="+30 minutes" +%M)"
			fi
		fi
	done
	
	echo '[FATAL ERROR] - $?'
	
	
#ERROR CODES
#E1 Can't Login to CyberGhost - Credentials not provided
#E2 VPN Connection Lost
#E3 Internet Connection Lost
	
